File: Introduction to Data Structures: Array and String/5. Conclusion/Rotate Array/RotateArray2.py
# second approach using reverse list

class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        def reverseList(l, r, lst):
            # reverse the whole list
            l, r = 0, len(lst)-1

            while l <= r:
                lst[l], lst[r] = lst[r], lst[l]
                l += 1
                r -= 1

        # handle when k > length nums
        if k > len(nums):
            k = k % len(nums)

        # reverse the whole list
        reverseList(0, len(nums)-1, nums)

        # reverse the first k elements
        reverseList(0, k-1, nums)

        # reverse the last remaining elements
        reverseList(k, len(nums)-1, nums)

File: Introduction to Data Structures: Array and String/5. Conclusion/Rotate Array/RotateArray2.java
class Solution {
    public void rotate(int[] nums, int k) {
        
        // handle if k is larger than the size of the input
        if (k > nums.length) {
            k = k % nums.length;   
        }
    
        // nums = [1,2,3,4,5,6,7], k = 3 Output: [5,6,7,1,2,3,4]
        // nums = [7,6,5,4,3,2,1]
        reverseArray(0, nums.length - 1, nums);
        
        // nums = [5,6,7,4,3,2,1]
        reverseArray(0, k - 1, nums);
        
        // nums = [5,6,7,1,2,3,4]
        reverseArray(k, nums.length - 1, nums);
    }
    
    // given the indices and array, it will reverse it in place
    private void reverseArray(int left, int right, int[]nums) {
        int temp;
        while (left <= right) {
            temp = nums[left];
            nums[left] = nums[right];
            nums[right] = temp;

            left++;
            right--;
        }
    }
}